Robert/addr memory leak

robert-zaremba opened the pull request:
## Description

In #8694 cache for `AccAddress.String()` was added as a simple map without control and estimates how that map will grow. This leads to unmanageable memory growth, which could put a node under a risk.

In this PR, 
+ we add a benchmark for the Address.String method.
+ we add an LRU cache instead of the map. 
+ remove deferred calls
+ optimize the `Address.Empty` function

Before:
```
BenchmarkAccAddressString-4   	 4533326	       268 ns/op	     144 B/op	       3 allocs/op
BenchmarkAccAddressString-4   	 4592979	       283 ns/op	     144 B/op	       3 allocs/op
```
After:
```
BenchmarkAccAddressString-4   	 4180477	       279 ns/op	     144 B/op	       3 allocs/op
BenchmarkAccAddressString-4   	 4633738	       274 ns/op	     144 B/op	       3 allocs/op
```

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

robert-zaremba commented:
BTW: we could add same thing for `ValAddress` and `ConsAddress`.
However, to be honest, I don't see much sense adding cache for `Address`. There are tonnes of other, better places for optimization without using a cache. Since it was added, I wanted to fix the memory issue.

alessio commented on types/address.go:
```diff
@@ -158,12 +159,7 @@ func (aa AccAddress) Equals(aa2 Address) bool {
-
-	aa2 := AccAddress{}
-	return bytes.Equal(aa.Bytes(), aa2.Bytes())
+	return aa == nil || len(aa) == 0
```
:+1:

alessio commented on types/address.go:
```diff
@@ -239,37 +235,33 @@ func (aa AccAddress) Bytes() []byte {
 	bech32PrefixAccAddr := GetConfig().GetBech32AccountAddrPrefix()
-
-	bech32Addr, err := bech32.ConvertAndEncode(bech32PrefixAccAddr, aa.Bytes())
+	bech32Addr, err := bech32.ConvertAndEncode(bech32PrefixAccAddr, aa)
```
Why?

robert-zaremba replied:
No need to call function. `aa` is bytes.

alessio commented on types/address_bench_test.go:
```diff
@@ -11,6 +11,18 @@ import (
+	for i := 0; i < b.N; i++ {
+		str = aa.String()
+	}
+	require.NotEmpty(b, str)
```
Shouldn't you move this up to after line 21?

robert-zaremba replied:
I want to benchmark `String`. It's enough if we check only once, no need to do that check in the loop every time.

alessio replied:
Well, it's a bit unnecessary then, isn't it? Not a big deal either way

odeke-em replied:
@robert-zaremba is on point here, just check only at the end, this is an idiomatic way of checking.

alessio commented on types/address.go:
```diff
@@ -239,37 +235,33 @@ func (aa AccAddress) Bytes() []byte {
-// in that pattern. See https://github.com/cosmos/cosmos-sdk/issues/8693.
 var addMu sync.Mutex
-var addrStrMemoize = make(map[string]string)
+var addrStrMemoize, _ = simplelru.NewLRU(1000, nil)
```
This initialization should be in an `init()` function, so that we can check the error returned and `panic()` if it is != `nil`.

robert-zaremba replied:
I was following the current design. That being said I agree with you - I will use `init`.

alessio reviewed (commented):
ACK on the concept from me

marbar3778 commented on types/address.go:
```diff
@@ -239,37 +235,33 @@ func (aa AccAddress) Bytes() []byte {
-// in that pattern. See https://github.com/cosmos/cosmos-sdk/issues/8693.
 var addMu sync.Mutex
-var addrStrMemoize = make(map[string]string)
+var addrStrMemoize, _ = simplelru.NewLRU(1000, nil)
```
1000 seems fairly small here. If we bump it to 60k (kind of reasonable) the memory footprint is 19.2 megabytes at with a size 32 bytes per address. I would feel comfortable with this size.

robert-zaremba replied:
the storage is address size + key size + hasha map residual space. I was thinking about number of address can fit in on block. Sure we can bump to bigger size. So how about 30k , to take into account the `key` storage?

robert-zaremba replied:
Updated to 30k

alessio reviewed (approved):
ACK

odeke-em commented on types/address.go:
```diff
@@ -237,39 +249,30 @@ func (aa AccAddress) Bytes() []byte {
-	if str, ok := addrStrMemoize[string(aa)]; ok {
-		return str
+func (aa AccAddress) String() string {
+	var key = string(aa)
```
This burns an allocation from []byte->str, we need to fix this and we have code for it :-)

robert-zaremba replied:
I was thinking about it, but I didn't want to introduce unsafe code here.
Nevertheless, as you see in the benchmark, the new code is doing slightly less allocations and has same speed.

robert-zaremba replied:
I created a PR for moving the unsafe convertion code to an internal `conv` package: https://github.com/cosmos/cosmos-sdk/pull/8733

odeke-em commented on types/address.go:
```diff
@@ -237,39 +249,30 @@ func (aa AccAddress) Bytes() []byte {
-	}()
 
 	if aa.Empty() {
+		addrMu.Lock()
+		addrCache.Add(key, "")
+		addrMu.Unlock()
```
There is no need to add this, I'd move all this into the defer as it was before.

robert-zaremba replied:
Defer is doing extra allocation ;) 
If we won't add a lock here, the race detector will complain.

odeke-em commented on types/address.go:
```diff
@@ -237,39 +249,30 @@ func (aa AccAddress) Bytes() []byte {
 	return aa
 }
 
-// AccAddress.String() is expensive and if unoptimized dominantly showed up in profiles,
-// yet has no mechanisms to trivially cache the result given that AccAddress is a []byte type.
-// With the string interning below, we are able to achieve zero cost allocations for string->[]byte
-// because the Go compiler recognizes the special case of map[string([]byte)] when used exactly
-// in that pattern. See https://github.com/cosmos/cosmos-sdk/issues/8693.
```
Please keep part of this explanation along with the issue number, as it will serve folks in the future well.

robert-zaremba replied:
But the `map` part is not relevant any more here - we are not using map.

odeke-em replied:
Yeah so pick out the relevant parts and keep the issue reference too :-)

robert-zaremba replied:
That was done before submitting the PR

odeke-em replied:
I replied because of of your reply :-) in https://github.com/cosmos/cosmos-sdk/pull/8717#discussion_r584717908

odeke-em commented on types/address_bench_test.go:
```diff
@@ -11,6 +11,18 @@ import (
+	pkBz := make([]byte, ed25519.PubKeySize)
+	pk := &ed25519.PubKey{Key: pkBz}
+	aa := pk.Address()
+	var str string
+	for i := 0; i < b.N; i++ {
+		str = aa.String()
+	}
+	require.NotEmpty(b, str)
```
To better get a good benchmark in which code won't be optimized away, use a global variable that's an interface e.g.
```go
var sink, revert interface{}
...
for i := 0; i < b.N; i++ {
   result := ...
   sink = result
}
if  sink == nil {
   b.Fatal("benchmark never ran")
}
sink = revert
```

robert-zaremba replied:
You mean a variable outside of any function? We are not using it in other benchmarks.

odeke-em replied:
Well just because it wasn't done correctly before doesn't mean we shouldn't do it correctly, I certainly have submitted many other benchmarks where we set to a global sink and check that.

robert-zaremba replied:
I checked it locally - it doesn't change anything. Are you sure we need to trick the benchmark code? 
My only objection is that we make the whole thing less readable and more complex without any observable effect.

odeke-em replied:
I am sure about code at times getting optimized away and that being the only way to block that. Humbly, for a bit of clarity, I work on the Go project, across almost all the packages and I am bringing this advice from the many years of working on benchmarking related code :-)

robert-zaremba replied:
I see. But why to obfuscate more if there is no difference (I verified it)? The `require.NoEmpty` already assures that value must not be ignored.

odeke-em reviewed (changes requested):
Thank you for this change @robert-zaremba! My apologies for the late reply, I was swamped with a couple of things.

A few things:
a) Please make the []byte->string conversion zero cost and not incur allocations
b) Please revert to using defers, the Go compiler made defer dirt cheap as of Go1.13 in 2019 so these are no longer required, please see https://go.googlesource.com/proposal/+/refs/heads/master/design/34481-opencoded-defers.md which was also publicized for that release
c) To get a really good reason to benchmark the LRU just remember that even if each Bech32 address is 90 characters at most as per https://wiki.trezor.io/Bech32#:~:text=A%20Bech32%20string%20is%20at,The%20human%2Dreadable%20part.&text=The%20separator%2C%20which%20is%20always%20%221%22, to consume 1GiB of RAM, that would need to process ~12M Bech32 addresses per `(1<<30)/90`, am not sure how often this will happen
d) the consideration of the LRU's contention should be checked and this can be done by writing a benchmark with say 1006 (>LRU_SIZE) bech32 values, run the benchmark before and after this change, and then run `benchstat before.txt after.txt`

Thank you!

robert-zaremba commented:
We can make more sophisticated benchmarks later (feel free to create a task for that). 
I will add the unsafe conversion - but I don't think this is needed (especially if we want to use `defer` - which will create a new function stack)

robert-zaremba commented:
1. Updated cache size
2. Moved Empty check before checking cache. Now Empty check is  very fast - so we don't need to cache it.
3. Applied same mechanism to other address types: `ValAddress` and `ConsAddress`.
4. Used unsafe bytes->string conversion.

odeke-em reviewed (commented):
Thank you for the next round of review, @robert-zaremba!
 
>We can make more sophisticated benchmarks later (feel free to create a task for that).

Am not sure what you mean by "sophisticated benchmarks", we need to get conclusive data on the performance and results. I did describe above how to do it. This change is adding in an LRU, replacing a global cache, whose size will only matter if we hit millions of accounts. If the performance of the LRU is worse than for the global map at scale, then we perhaps have other problems.

For benchmarks, after you run them before and after, please use `benchstat`https://pkg.go.dev/golang.org/x/tools/cmd/benchstat to give a summary of the comparisons otherwise humans can't tell what changed easily. Also to ensure you get good consistent runs, please use `-count=5` or even more, to like `-count=20`.

The benchmark that I had suggested earlier on will give us conclusive results on the impacts and changes plus what performance improvements or adverse effects that this change brings.

Thanks for the PR.

robert-zaremba commented on types/address.go:
```diff
@@ -237,40 +259,18 @@ func (aa AccAddress) Bytes() []byte {
 	}
-
-	return bech32Addr
+	return cacheBech32Addr(GetConfig().GetBech32AccountAddrPrefix(),
```
I don't think so - the config must be loaded dynamically. It's done in many places like that. Refactoring that is a future task, and should consider other places as well.

odeke-em replied:
Interesting about the configuration being reloaded dynamically, alright in that case please dismiss my comment.

robert-zaremba commented on types/address.go:
```diff
@@ -730,3 +716,14 @@ func addressBytesFromHexString(address string) ([]byte, error) {
 	return hex.DecodeString(address)
 }
+
+func cacheBech32Addr(prefix string, addr []byte, m sync.Locker, cache *simplelru.LRU, cacheKey string) string {
```
According to our linter rules, interface should be used here -> so `sync.Locker`

odeke-em replied:
Interesting! In that case then please dismiss my comment.

robert-zaremba commented:
@odeke-em  - adding more benchmarks is going out of scope and we should have a good to have a real world scenario here. We can do it in other issue. 

The intention of this PR was to remove the memory leak. I think it's better to have this update then rolling back the cache. I suggest to create a new issue with further benchmarks.

I added one more, non trivial address to the benchmark.

amaurym commented on types/address.go:
```diff
@@ -72,6 +74,31 @@ const (
+
+func init() {
+	var err error
+	if accAddrCache, err = simplelru.NewLRU(25000, nil); err != nil {
```
would be good to add some comments on how these numbers are chosen (copy paste what has been said in the comments here)

robert-zaremba replied:
sure, will add it. There was not much analysis behind it.

amaurym reviewed (approved):
utACK

robert-zaremba commented:
@odeke-em  your lock on this PR is still active. Please comment if there is something wrong in this PR. Adding more benchmarks is, imho, out of scope - we would need to go into use-cases  / scenario bench-marking to bring bore valuable information.

odeke-em commented:
Thanks for the ping, Robert! I’ll remove my review to avoid blocking you but I definitely think that without a proper and conclusive benchmark, we might not get the right results, as the point of using an LRU was to counteract hogging up memory that won’t get evicted. If the cost of using an LRU is worse that’s a problem.

robert-zaremba commented:
>  If the cost of using an LRU is worse that’s a problem.

The benchmark (which extensively uses cache) shows that's not worse. Moreover the new code uses less allocations - this is due to avoiding `defer` and optimizing "empty" case.

colin-axner commented:
The race-detector tests appears to begin failing with the merge of this pr. Maybe it is due to something else, but thought I'd mention it here

robert-zaremba commented:
@colin-axner  What exactly and Where do you see the test failing?
Here, in the merge checks nothing was failing.
When running `make test-race` I see the following things failing:

+ FAIL: TestListSnapshots (137.46s)
+ FAIL	github.com/cosmos/cosmos-sdk/baseapp	468.134s

It's hard to digest the output of `maket test-race`. There are tonnes of lines related to `leveldb`....

The only interesting thing I found is the following warning:
```
WARNING: DATA RACE
Write at 0x00c000fbf470 by goroutine 43:
  container/list.(*List).move()
      /usr/lib/go-1.15/src/container/list/list.go:123 +0x28a
  container/list.(*List).MoveToFront()
      /usr/lib/go-1.15/src/container/list/list.go:188 +0x23f
  github.com/hashicorp/golang-lru/simplelru.(*LRU).Get()
      /home/robert/go/pkg/mod/github.com/hashicorp/golang-lru@v0.5.4/simplelru/lru.go:75 +0xed
  github.com/cosmos/cosmos-sdk/types.AccAddress.String()
      /home/robert/go/src/github.com/cosmos/cosmos-sdk/types/address.go:271 +0x152
```
This access is OK - because we are fine with a concurrent read, even if it's done during the read - we write data only once, and it never depends on what we read. I can add a mutex there if we want to mute that warning. What do you think?

colin-axner commented:
Ah, I got a little confused by not looking at the full logs

I also see this (and similar errors in a couple other places):
```
2021-03-03T03:52:25.9594698Z --- FAIL: TestIntegrationTestSuite (102.02s)
2021-03-03T03:52:25.9595884Z     --- FAIL: TestIntegrationTestSuite/TestCLIMultisign (11.95s)
2021-03-03T03:52:25.9597152Z     --- FAIL: TestIntegrationTestSuite/TestCLISignBatch (4.64s)
2021-03-03T03:52:25.9598103Z     testing.go:1038: race detected during execution of test
2021-03-03T03:52:25.9598655Z FAIL
2021-03-03T03:52:25.9599405Z FAIL	github.com/cosmos/cosmos-sdk/x/auth/client/cli	102.710s
```

My main concern is having the race-detector become stuck on always failing. I never check the logs, so I imagine it could be dangerous to have a test which always fails since if someone does introduce a race condition, it could go unnoticed

I have little context on this pr, so will differ to the decisions of others here. Maybe the other `test-race` tests are a good enough to ensure race conditions don't get introduced

robert-zaremba commented:
Yeah, the logs are huuuuge. I will add a check and let's see what will happen.

alessio commented:
Oh wow - why was mergify allowed to merge something whose tests are not all green in the first place?

@marbar3778 can we do something to prevent this from happening again?

alessio commented:
I think I know why. The `race-detector-report` job was not mandatory for PRs to be merged in.

colin-axner commented:
It should be noted that I don't think any pr's can be merged until the race detector is fixed

alessio commented:
> It should be noted that I don't think any pr's can be merged until the race detector is fixed

The branch configuration is now fixed, so PRs will be blocked until the race is fixed. Does that work for you?

colin-axner commented:
It is fine with me, but it might be blocking a 0.42.0 final release. 

I'm hoping to remove IBC #8735 this week (so it is included in 0.42). That pr is just waiting on approval of #8624 which I guess now will wait on a race-detector fix

Overall though, I am most certainly in favor of fixing tests before merging more pr's and making the problem worse. Just hope we can do so soon

robert-zaremba commented:
PR: https://github.com/cosmos/cosmos-sdk/pull/8767
